src/slice.rs

```diff
@@ -623,3 +623,273 @@ macro_rules! s(
+    _view: &ArrayViewMut<'a, A, D>
+) -> PhantomData<&'a mut A> {
+    PhantomData
+}
+
+/// Derefs the raw mutable view into a view, using the given lifetime.
+#[doc(hidden)]
+pub unsafe fn deref_raw_view_mut_into_view_with_life<'a, A, D: Dimension>(
+    raw: RawArrayViewMut<A, D>,
+    _life: PhantomData<&'a mut A>,
+) -> ArrayView<'a, A, D> {
+    raw.deref_into_view()
+}
+
+/// Derefs the raw mutable view into a mutable view, using the given lifetime.
+#[doc(hidden)]
+pub unsafe fn deref_raw_view_mut_into_view_mut_with_life<'a, A, D: Dimension>(
+    raw: RawArrayViewMut<A, D>,
+    _life: PhantomData<&'a mut A>,
+) -> ArrayViewMut<'a, A, D> {
+    raw.deref_into_view_mut()
+}
+
+/// Take multiple slices simultaneously.
+///
+/// This macro makes it possible to take multiple slices of the same array, as
+/// long as Rust's aliasing rules are followed for *elements* in the slices.
+/// For example, it's possible to take two disjoint, mutable slices of an
+/// array, with one referencing the even-index elements and the other
+/// referencing the odd-index elements. If you tried to achieve this by calling
+/// `.slice_mut()` twice, the borrow checker would complain about mutably
+/// borrowing the array twice (even though it's safe as long as the slices are
+/// disjoint).
+///
+/// The syntax is `multislice!(` *expression, (pattern [, pattern [, …]])* `)`,
+/// where *expression* evaluates to an `ArrayBase<S, D>` where `S: DataMut`,
+/// and `pattern` is one of the following:
+///
+/// * `mut expr`: creates an `ArrayViewMut`, where `expr` evaluates to a
+///   `&SliceInfo` instance used to slice the array.
+/// * `expr`: creates an `ArrayView`, where `expr` evaluates to a `&SliceInfo`
+///   instance used to slice the array.
+///
+/// **Note** that this macro always mutably borrows the array even if there are
+/// no `mut` patterns. If all you want to do is take read-only slices, you
+/// don't need `multislice!()`; just call
+/// [`.slice()`](struct.ArrayBase.html#method.slice) multiple times instead.
+///
+/// `multislice!()` evaluates to a tuple of `ArrayView` and/or `ArrayViewMut`
+/// instances. It checks Rust's aliasing rules:
+///
+/// * An `ArrayViewMut` and `ArrayView` cannot reference the same element.
+/// * Two `ArrayViewMut` cannot reference the same element.
+/// * Two `ArrayView` can reference the same element.
+///
+/// **Panics** at runtime if any of the aliasing rules is violated.
+///
+/// See also [*Slicing*](struct.ArrayBase.html#slicing).
+///
+/// # Examples
+///
+/// In this example, there are two overlapping read-only slices, and two
+/// disjoint mutable slices. Neither of the mutable slices intersects any of
+/// the other slices.
+///
+/// ```
+/// extern crate ndarray;
+///
+/// use ndarray::multislice;
+/// use ndarray::prelude::*;
+///
+/// # fn main() {
+/// let mut arr = Array1::from_iter(0..12);
+/// let (a, b, c, d) = multislice!(arr, (s![0..5], mut s![6..;2], s![1..6], mut s![7..;2]));
```

Comment:
That's a good point. I wrote it that way because it was the most obvious way to allow the user to supply a `&SliceInfo` instance instead of always requiring them to use `s!`-style notation. However, it would be possible to use pattern-matching to differentiate between `s!`-style notation and an expression that evaluates to `&SliceInfo`, so the `s!` isn't strictly necessary. That said, I think that we should keep it as-is for the following reasons:

* It's easier to read `multislice!(arr, (s![0..5], mut s![6..;2]))` at a glance than `multislice!(arr, ([0..5], mut [6..;2]))` if you have syntax highlighting because the `s!` is shown in a different color, which makes it really easy to see the grouping of the slicing info.
* When the user writes `multislice!(arr, (s![0..5], mut s![6..;2]))`, it's clear what the `s!` means because the `s!` macro is used fairly often. Without the `s!`, it's not quite as obvious.
* The `multislice!` implementation is a little simpler without having to provide special parsing for `s!`-style expressions.

Reply:
I'm curious, is it technically possible to use this syntax: `multislice!(arr, [0..5], mut [6..;2])`?

Reply:
Yes; we can just add rules that apply the `s![]` macro, matching on `[$($info:tt)*]`, above the rules matching on `$info:expr`:

```diff
diff --git a/src/slice.rs b/src/slice.rs
index 0d358f1..a06a8fd 100644
--- a/src/slice.rs
+++ b/src/slice.rs
@@ -770,6 +770,108 @@ macro_rules! multislice(
             )
         }
     };
+    // Parse last slice (mutable), no trailing comma, applying `s![]` macro.
+    (
+        @parse $view:expr, $life:expr,
+        ($($sliced:tt)*),
+        ($($mut_info:tt)*),
+        ($($immut_info:tt)*),
+        (mut [$($info:tt)*])
+    ) => {
+        // Apply `s![]` macro to info.
+        $crate::multislice!(
+            @parse $view, $life,
+            ($($sliced)*),
+            ($($mut_info)*),
+            ($($immut_info)*),
+            (mut s![$($info)*],)
+        )
+    };
+    // Parse last slice (read-only), no trailing comma, applying `s![]` macro.
+    (
+        @parse $view:expr, $life:expr,
+        ($($sliced:tt)*),
+        ($($mut_info:tt)*),
+        ($($immut_info:tt)*),
+        ([$($info:tt)*])
+    ) => {
+        // Apply `s![]` macro to info.
+        $crate::multislice!(
+            @parse $view, $life,
+            ($($sliced)*),
+            ($($mut_info)*),
+            ($($immut_info)*),
+            (s![$($info)*],)
+        )
+    };
+    // Parse last slice (mutable), with trailing comma, applying `s![]` macro.
+    (
+        @parse $view:expr, $life:expr,
+        ($($sliced:tt)*),
+        ($($mut_info:tt)*),
+        ($($immut_info:tt)*),
+        (mut [$($info:tt)*],)
+    ) => {
+        // Apply `s![]` macro to info.
+        $crate::multislice!(
+            @parse $view, $life,
+            ($($sliced)*),
+            ($($mut_info)*),
+            ($($immut_info)*),
+            (mut s![$($info)*],)
+        )
+    };
+    // Parse last slice (read-only), with trailing comma, applying `s![]` macro.
+    (
+        @parse $view:expr, $life:expr,
+        ($($sliced:tt)*),
+        ($($mut_info:tt)*),
+        ($($immut_info:tt)*),
+        ([$($info:tt)*],)
+    ) => {
+        // Apply `s![]` macro to info.
+        $crate::multislice!(
+            @parse $view, $life,
+            ($($sliced)*),
+            ($($mut_info)*),
+            ($($immut_info)*),
+            (s![$($info)*],)
+        )
+    };
+    // Parse a mutable slice, applying `s![]` macro.
+    (
+        @parse $view:expr, $life:expr,
+        ($($sliced:tt)*),
+        ($($mut_info:tt)*),
+        ($($immut_info:tt)*),
+        (mut [$($info:tt)*], $($t:tt)*)
+    ) => {
+        // Apply `s![]` macro to info.
+        $crate::multislice!(
+            @parse $view, $life,
+            ($($sliced)*),
+            ($($mut_info)*),
+            ($($immut_info)*),
+            (mut s![$($info)*], $($t)*)
+        )
+    };
+    // Parse a read-only slice, applying `s![]` macro.
+    (
+        @parse $view:expr, $life:expr,
+        ($($sliced:tt)*),
+        ($($mut_info:tt)*),
+        ($($immut_info:tt)*),
+        ([$($info:tt)*], $($t:tt)*)
+    ) => {
+        // Apply `s![]` macro to info.
+        $crate::multislice!(
+            @parse $view, $life,
+            ($($sliced)*),
+            ($($mut_info)*),
+            ($($immut_info)*),
+            (s![$($info)*], $($t)*)
+        )
+    };
     // Parse last slice (mutable), no trailing comma.
     (
         @parse $view:expr, $life:expr,
```

This change would be backwards compatible if we want to add it in the future.

Reply:
Just counting in syntactic noise, it seems like a big win to drop both the extra () around the slices and each `s!`, I would go for that right away.  The `s!` would keep working, because you've designed it so we can insert arbitrary expressions for each axis, right?  You could do (?)

```rust
let ax1 = s![1..];
let ax2 = s![2..];
multislice!(arr, ax1, ax2);
```

I guess it would count as us breaking array expressions in those locations, but there is no reason to use those there at this point.

Reply:
Oh, I didn't notice the dropped () in your earlier comment. Yeah, we can drop the () too. With the extra () dropped, I think it's fine to drop the `s!` too. (With the (), it looks like a tuple, which makes me want to require the extra `s!` so that the elements are expressions. Without the (), the parameters are clearly parameters to the macro and not part of a tuple, so IMO it's fine to do things like interpreting them as `s!`-style syntax without requiring the `s!`.)

I'll update the PR to drop the extra () and `s!`.

> The `s!` would keep working, because you've designed it so we can insert arbitrary expressions for each output, right?

Yes, normal expressions that evaluate to `&SliceInfo` will still work. (They just have a lower precedence than `[$($info:tt)*]`.)

> I guess it would count as us breaking array expressions (like `[1,2,3]`) in those locations, but there is no reason to use those there at this point.

Yeah, these don't matter right now, because the slicing expressions must have type `&SliceInfo`, and an expression of the form `[$($info:tt)*]` is always a fixed-length array, not a `&SliceInfo`. If we want to add support for slicing using fixed-length array expressions in the future (e.g. `[1, 2, 3]` and `[0..5, 3..7, 2..8]` for `Ix3`) that don't use the `s!` macro machinery, we can do so by adding more rules with even higher precedence. (They'd match on e.g. `[$($info:expr),*]`.)

One interesting thing to note is that we could implement `multislice!` as a method on `ArrayBase`, instead of a macro, that could be called like this:

```rust
arr.multislice((s![0..5], Mut(s![6..;2])))
// or
arr.multislice(ss!([0..5], mut [6..;2]))
```

but that would work up to only a limited tuple length (however long we'd decide to support), like `Zip` works up to only 6 producers.

Reply:
As a method sounds cool but the syntax doesn't work quite so well there, so it doesn't seem as good. Hopefully multislice as a macro has its own attraction by virtue of its functionality so that users find it.

Reply:
I actually think `arr.multislice(ss!([0..5], mut [6..;2]))` is quite nice, but it would be more complex to implement than the `multislice!` macro.

I've updated the PR to drop the extra () and `s!`.